src/Mvc/Mvc.Core/src/Routing/DynamicControllerEndpointMatcherPolicy.cs

```diff
@@ -97,13 +97,17 @@ public async Task ApplyAsync(HttpContext httpContext, CandidateSet candidates)
                 // no realistic way this could happen.
                 var dynamicControllerMetadata = endpoint.Metadata.GetMetadata<DynamicControllerMetadata>();
                 var transformerMetadata = endpoint.Metadata.GetMetadata<DynamicControllerRouteValueTransformerMetadata>();
+
+                DynamicRouteValueTransformer transformer = null;
                 if (dynamicControllerMetadata != null)
                 {
                     dynamicValues = dynamicControllerMetadata.Values;
                 }
                 else if (transformerMetadata != null)
                 {
-                    var transformer = (DynamicRouteValueTransformer)httpContext.RequestServices.GetRequiredService(transformerMetadata.SelectorType);
+                    transformer = (DynamicRouteValueTransformer)httpContext.RequestServices.GetRequiredService(transformerMetadata.SelectorType);
```

Comment:
>         /// The instance of <typeparamref name="TTransformer"/> will be retrieved from the dependency injection container.
>         /// Register <typeparamref name="TTransformer"/> as transient in <c>ConfigureServices</c>. Using the transient lifetime
>         /// is required when using <paramref name="state" />.

Are there any guardrails that can be put in place to stop developers using state with DynamicRouteValueTransformer's that aren't registered as non-transient? It seems like an easy mistake to make that would introduce hard to debug concurrency problems.

The only thing I can think of is validating that `transformer.State` is null when an instance is returned from `GetRequiredService`. However someone could have decided to set the state property in the transformer constructor which would trigger this error.

Reply:
That's a fair point, but I don't think we have a good way to do it, as you mention. I think covering it in docs is a reasonable first approach. If we detect lots of people falling into this trap, then we can evaluate additional protections. If it were a singleton, one thing we could do is resolve it twice and check for `Object.ReferenceEquals(a,b) == true` but that's something I wouldn't do on the main code path. Another option would be to add an analyzer, but I don't think we need it.

From what I've seen this is a pretty advanced feature and the type of users that rely on it are familiar with the internals, so it is unlikely they make such mistakes.

Reply:
What do you think of checking `transformer.State` is always null? Having that as requirement is something that would need to be done now. Requiring it later would be a breaking change.

```cs
transformer = (DynamicRouteValueTransformer)httpContext.RequestServices.GetRequiredService(transformerMetadata.SelectorType);
if (transformer.State != null)
{
    throw new InvalidOperationException($"Transformer {transformerMetadata.SelectorType.Name} was retrieved from dependency injection with a state value. State can only be specified using the MapDynamicControllerRoute method's state argument and a transient transformer. {transformerMetadata.SelectorType.Name} may not be registered as transient in dependency injection");
}
```

Reply:
@JamesNK That's fair. I think its trying to prevent a problem before we know e have it, but it is cheap and we can relax this condition later if we find out for some reasons there are valid scenarios for it.

Reply:
What do you think of checking `transformer.State` is always null? Having that as requirement is something that would need to be done now. Requiring it later would be a breaking change.

I think this could work:

```cs
transformer = (DynamicRouteValueTransformer)httpContext.RequestServices.GetRequiredService(transformerMetadata.SelectorType);
if (transformer.State != null)
{
    throw new InvalidOperationException($"Transformer {transformerMetadata.SelectorType.Name} was retrieved from dependency injection with a state value. " +
        "State can only be specified when the dynamic route is mapped using the state argument together with transient lifetime transformer. " +
        $"Ensure that {transformerMetadata.SelectorType.Name} doesn't set its own state and that the transformer is registered with a transient lifetime in dependency injection.");
}
```